### ISSUE TITLE
Inner-query with byte[] FK produces wrong result with EFCore 2.1.x

### ISSUE THREAD

**Issue by AtsushiSuzuki** [aspnet/EntityFrameworkCore#12748]
After upgrading EFCore to 2.1.x, specific kind of query produces wrong result.

### Steps to reproduce
I have following models
```c#
    public class Blog
    {
        [Key]
        public byte[] Name { get; set; }
        public List<Comment> Comments { get; set; }
    }

    public class Comment
    {
        public int Id { get; set; }
        public byte[] BlogName { get; set; }
        public Blog Blog { get; set; }
    }
```

and following query.
```c#
                var query = from blog in ctx.Blogs
                            select new
                            {
                                blog.Name,
                                Comments = blog.Comments.Select(u => new
                                {
                                    Id = u.Id,
                                }).ToArray(),
                            };
```

complete example is at https://gitlab.com/suzuki_ds/efcore_test .
Test1 in UnitTest1.cs describes this issue.

#### Expected behavior
Result items should have `Comments` collection.

#### Actual behavior
Result items have `Comments` collection with no element. (size == 0)

This problem will happen only on EFCore 2.1.x, not on 2.0.x.

### Further technical details
EF Core version: Microsoft.AspNetCore.All 2.1.2
Database Provider:Microsoft.EntityFrameworkCore.Sqlite 2.1.1
Operating system: Windows 10 Pro 1803
IDE: Visual Studio 2017 15.7.5

**Comment by maumar** [comment]
We generate the following query plan:

```
(QueryContext queryContext) => IEnumerable<<>f__AnonymousType0<byte[], <>f__AnonymousType1<int>[]>> _InterceptExceptions(
|__ source: IEnumerable<<>f__AnonymousType0<byte[], <>f__AnonymousType1<int>[]>> _ShapedQuery(
|   |__ queryContext: queryContext, 
|   |__ shaperCommandContext: SelectExpression: 
|   |       SELECT [blog].[Name]
|   |       FROM [Blogs] AS [blog]
|   |       ORDER BY [blog].[Name], 
|   |__ shaper: (QueryContext queryContext | ValueBuffer blog) => new { 
|           byte[] TryReadValue(blog, 0, Blog.Name), 
|           <>f__AnonymousType1<int>[] ToArray(List<<>f__AnonymousType1<int>> queryContext.QueryBuffer.CorrelateSubquery(
|           |__ correlatedCollectionId: 0, 
|           |__ navigation: Blog.Comments, 
|           |__ resultCollectionFactory: (INavigation n) => new List<<>f__AnonymousType1<int>>(), 
|           |__ outerKey: new MaterializedAnonymousObject(new object[]{ (object)!(blog.IsEmpty) ? byte[] TryReadValue(blog, 0, Blog.Name) : default(byte[]) }), 
|           |__ tracking: False, 
|           |__ correlatedCollectionFactory: () => IEnumerable<Tuple<<>f__AnonymousType1<int>, MaterializedAnonymousObject, MaterializedAnonymousObject>> _ShapedQuery(
|           |   |__ queryContext: queryContext, 
|           |   |__ shaperCommandContext: SelectExpression: 
|           |   |       SELECT [t].[Name], [blog.Comments].[Id], [blog.Comments].[BlogName]
|           |   |       FROM [Comments] AS [blog.Comments]
|           |   |       INNER JOIN (
|           |   |           SELECT [blog0].[Name]
|           |   |           FROM [Blogs] AS [blog0]
|           |   |       ) AS [t] ON [blog.Comments].[BlogName] = [t].[Name]
|           |   |       ORDER BY [t].[Name], 
|           |   |__ shaper: (QueryContext queryContext | TransparentIdentifier<ValueBuffer, MaterializedAnonymousObject> t0) => new Tuple<<>f__AnonymousType1<int>, MaterializedAnonymousObject, MaterializedAnonymousObject>(
|           |           new { int TryReadValue(t0.Outer, 1, Comment.Id) }, 
|           |           new MaterializedAnonymousObject(new object[]{ (object)!(t0.Outer.IsEmpty) ? byte[] TryReadValue(t0.Outer, 2, Comment.BlogName) : default(byte[]) }), 
|           |           new MaterializedAnonymousObject(new object[]{ (object)(object)object TryReadValue(t0.Outer, 0, Blog.Name) })
|           |       )), 
|           |__ correlationPredicate: (MaterializedAnonymousObject o | MaterializedAnonymousObject i) => object i.GetValue(0) == default(object) || object o.GetValue(0) == default(object) ? False : (byte[])object o.GetValue(0) == (byte[])object i.GetValue(0)))
|        }), 
|__ contextType: Repro12748.MyContext, 
|__ logger: DiagnosticsLogger<Query>, 
|__ queryContext: Unhandled parameter: queryContext)
```

problem is that corelationPredicate does regular comparison for byte[], we should be performing structural comparison on those values instead.

**Comment by maumar** [comment]
fixed in e38ead5e17ff8afda8a69c9879cf7b6b00b57a47
